stdlib/json.gr

```diff
@@ -0,0 +1,2155 @@
+/**
+ * JSON (JavaScript Object Notation) parsing, printing, and access utilities.
+ */
+
+module Json
+
+include "runtime/bigint" as BI
+include "runtime/dataStructures"
+include "runtime/numbers"
+include "runtime/numberUtils"
+include "runtime/string" as RunTimeString
+include "runtime/unsafe/memory"
+include "runtime/unsafe/memory"
+include "runtime/unsafe/tags"
+include "runtime/unsafe/wasmi32"
+include "runtime/unsafe/wasmi64"
+include "runtime/unsafe/wasmf32"
+include "runtime/unsafe/wasmf64"
+include "buffer"
+include "char"
+include "string"
+include "list"
+include "number"
+include "option"
+include "uint8"
+from RunTimeString use { toString as runtimeToString, getCodePoint }
+from Numbers use { coerceNumberToWasmI32 }
+from DataStructures use { tagSimpleNumber, newFloat64, untagSimpleNumber }
+
+// Primitive offsets
+// TODO(703): Get these offsets from the runtime
+@unsafe
+let _INT64_BOXED_VALUE_OFFSET = 8n
+@unsafe
+let _Float64_BOXED_VALUE_OFFSET = 8n
+
+/**
+ * Data structure representing the result of parsing and the input of printing
+ * JSON.
+ *
+ * For example this object
+ * ```grain
+ * JsonObject([
+ *   ("currency", JsonString("€")),
+ *   ("price", JsonNumber(99.99)),
+ * ])
+ * ```
+ * is equivalent to the following JSON:
+ * ```json
+ * {"currency":"€","price":99.99}
+ * ```
+ *
+ * This data structure is semantically equivalent to the JSON format allowing
+ * mostly lossless round trips of printing and parsing. Exceptions to this are
+ * whitespace, multiple ways JSON allows escaping characters in strings, and
+ * some edge cases related to Grain's `Number` type.
```

Comment:
I feel like this comment might add confusion for users. might be better if this comment was moved out of the grain doc and into an internal comment.

as per the number type, yes it is something like a big float would get rounded which doesn't semantically follow the spec (mainly because the spec only covers the syntax and not actual parsing, at least from what i understand)

Reply:
This is true—there are some edge cases, but that's also true for every JSON implementation ever. I don't think we should have this second section at all, especially since it's not actually related to the enum itself.

Reply:
I removed the `Exception` part I feel like the part about it being semantically equivalent is relevant

Reply:
I think that part is obvious—it's called JSON because it's representing a JSON document.

Reply:
Fair enough, I removed the entire thing.